bugsnag-android-core/src/main/java/com/bugsnag/android/JournaledStateObserver.kt

```diff
@@ -0,0 +1,171 @@
+            }
+            is StateEvent.StartSession -> {
+                handleStartSessionEvent(event)
+            }
+            is StateEvent.UpdateContext -> {
+                addJournalCommand(JournalKeys.pathContext, event.context)
+            }
+            is StateEvent.UpdateInForeground -> {
+                addJournalCommand(JournalKeys.pathAppInForeground, event.inForeground)
+                addJournalCommand(JournalKeys.pathMetadataAppActiveScreen, event.contextActivity)
+            }
+            is StateEvent.UpdateLastRunInfo -> {
+                // This event doesn't seem to actually be used.
+            }
+            is StateEvent.UpdateIsLaunching -> {
+                addJournalCommand(JournalKeys.pathAppIsLaunching, event.isLaunching)
+            }
+            is StateEvent.UpdateOrientation -> {
+                addJournalCommand(JournalKeys.pathDeviceOrientation, event.orientation)
+            }
+            is StateEvent.UpdateUser -> {
+                handleUpdateUserEvent(event)
+            }
+            is StateEvent.UpdateMemoryTrimEvent -> {
+                addJournalCommand(JournalKeys.pathMetadataAppLowMemory, event.isLowMemory)
+            }
+        }
+    }
+
+    private fun handleUpdateUserEvent(event: StateEvent.UpdateUser) {
+        addJournalCommand(
+            JournalKeys.pathUser,
+            mapOf(
+                JournalKeys.keyName to event.user.name,
+                JournalKeys.keyId to event.user.id,
+                JournalKeys.keyEmail to event.user.email
+            )
+        )
+    }
+
+    private fun handleAddBreadcrumbEvent(event: StateEvent.AddBreadcrumb) {
+        addJournalCommand(
+            "${JournalKeys.pathBreadcrumbs}.",
+            mapOf(
+                JournalKeys.keyMetadata to event.metadata,
+                JournalKeys.keyName to event.message,
+                JournalKeys.keyTimestamp to event.timestamp,
+                JournalKeys.keyType to event.type.toString()
+            )
+        )
+    }
+
+    private fun handleStartSessionEvent(event: StateEvent.StartSession) {
+        addJournalCommand(
+            JournalKeys.pathSession,
+            mapOf(
+                JournalKeys.keyId to event.id,
+                JournalKeys.keyStartedAt to event.startedAt,
+                JournalKeys.keyEvents to mapOf(
+                    JournalKeys.keyHandled to event.handledCount,
+                    JournalKeys.keyUnhandled to event.unhandledCount
+                )
+            )
+        )
+    }
+
+    private fun handleInstallEvent(event: StateEvent.Install) {
+        addJournalCommand(JournalKeys.pathApiKey, event.apiKey)
+        addJournalCommand(JournalKeys.pathContext, client.context)
+
+        val appDataCollector = client.getAppDataCollector()
+        val appData = appDataCollector.generateAppWithState()
+        addJournalCommand(
+            JournalKeys.pathApp,
+            mapOf(
```

Comment:
I'm 90% sure that another function exists in the codebase that already coerces app/device into a map, it may be worth searching for it to reuse if possible